### DISCUSSION TITLE
Node JS + Sequelize - Async Authentication Issues

### DISCUSSION THREAD

**Discussion by Jpm91297** [sequelize/sequelize#16345] [Help & Questions]
I am running a Node.Js app as an Azure Web App and am trying to implement Azure Secrets as a way to store the password to my SQL Server. When initializing my DB using Sequelize, I am making use of the beforeConnect() hook in order to asynchronously pull in the password from a helper function. Because of pooling, beforeConnect() is running every 10 seconds. In order to prevent repetitive calls to pull in the Azure Secret, I declared the password variable outside of the helper function and am only pulling it if is not there. The issue with this is that if the password changes, it will not refresh that variable unless I were to restart the application. I tried adding a parameter that would force the function to pull the new value if an error is thrown from DB.authenticate(), however I learned that this function only runs once when the server is first started. Any advice is greatly appreciated.

DB.js
```

const Sequelize = require('sequelize');
const Config = require('../Config/config');
const keyVaultHelperObj = require('../Config/keyVaultHelper')

let configSettings = {
    ...
}

const DB = new Sequelize(Config.ih3Database, Config.userName, '', configSettings);
DB.beforeConnect(async config => {
    config.password = await keyVaultHelperObj.getSQLPassword();
});

DB.authenticate().then(() => {
    console.log(`Connection to ${Config.ih3Database} established successfully.`)
}).catch(async err => {
    console.log(err)
    await keyVaultHelperObj.getSQLPassword('force')
})

module.exports = DB;
```

Helper Function:

```
const { DefaultAzureCredential } = require("@azure/identity");
const { SecretClient } = require("@azure/keyvault-secrets");

const credential = new DefaultAzureCredential();
let sqlPassword = null;

const keyVaultHelperObj = {
    getSQLPassword: async (force = false) => {
        if (!sqlPassword || force) {
            if (force) {
                console.log('Pulling new password....')
            }
            const vaultName = 'exampleVault';
            const url = `https://${vaultName}.vault.azure.net`;
            const client = new SecretClient(url, credential, { serviceVersion: '7.0' });
            let { value } = await client.getSecret('exampleSecret');
            console.log(value, '<=====');
            sqlPassword = value;
        }

        return sqlPassword;
    }
}

module.exports = keyVaultHelperObj;

```

**Comment by rajvansia** [comment]
@Jpm91297 thanks for sharing I am running into the same issue. I was looking at beforePoolAquire

**Comment by lwileczek** [comment]
Here is code from their v7 alpha, it appears this hook may not allow modifications to the config if the dialect is not postgresql.  I cannot find this same unit test for their v6. 

https://github.com/sequelize/sequelize/blob/1cd6981c38a99a465da7ffd0adf71494dfda9ec8/packages/core/test/unit/pool.test.ts#L73
